File: leetcode-04/solutions/python3/0954.py
class Solution:
    def canReorderDoubled(self, A: List[int]) -> bool:
        dict = collections.Counter(A)

        for key in sorted(dict, key=abs):
            if dict[key] > dict[2 * key]:
                return False
            dict[2 * key] -= dict[key]

        return True

File: leetcode-04/solutions/cpp/0954.cpp
class Solution {
 public:
  bool canReorderDoubled(vector<int>& A) {
    unordered_map<int, int> map;
    for (int a : A) map[a]++;

    vector<int> keys;

    for (auto& [key, _] : map) keys.push_back(key);

    sort(keys.begin(), keys.end(),
         [](int a, int b) { return abs(a) < abs(b); });

    for (int key : keys) {
      if (map[key] > map[2 * key]) return false;
      map[2 * key] -= map[key];
    }

    return true;
  }
};